Tpetra: Bad interaction with Kokkos ETI with XL compiler

## Bug Report
@trilinos/tpetra
@jhux2 
@brian-kelley 

### Description
SPARC is failing to link on vortex pwr9 seemingly because of the change in TpetraExt_MatrixMatrix_def.hpp from 75333e4ed. The XL compiler is the only place it has appeared. The error is as follows: 

```
/projects/atdm_devops/trilinos_installs/2020-12-02/ats2-xl-2020.03.18_spmpi-rolling_serial_static_opt/lib/libtpetraext.a(TpetraExt_MatrixMatrix.cpp.o):(.toc+0x68): undefined reference to `KokkosSparse::Experimental::SortedCountEntries<unsigned long, int, Kokkos::View<unsigned long const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<0u> >, Kokkos::View<unsigned long const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<0u> >, Kokkos::View<int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<0u> >, Kokkos::View<int const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<0u> >, Kokkos::View<unsigned long*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<0u> > >::ORDINAL_MAX'
```

Please confirm that the deduced template arguments are the intended template arguments when using `GO=long long` and `LO = int`.

--- a/packages/kokkos-kernels/src/sparse/KokkosSparse_spadd.hpp
+++ b/packages/kokkos-kernels/src/sparse/KokkosSparse_spadd.hpp
@@ -47,7 +47,7 @@
 
 #include "KokkosKernels_Handle.hpp"
 #include "KokkosKernels_Sorting.hpp"
-#include <limits>
+#include "Kokkos_ArithTraits.hpp"
 
 namespace KokkosSparse {
 namespace Experimental {
@@ -86,10 +86,10 @@ struct SortedCountEntries {
         Bcolinds(Bcolinds_),
         Crowcounts(Crowcounts_) {}
 
-  static constexpr ordinal_type ORDINAL_MAX = std::numeric_limits<ordinal_type>::max();
-
   KOKKOS_INLINE_FUNCTION void operator()(const ordinal_type i) const
   {
+    const ordinal_type ORDINAL_MAX = Kokkos::ArithTraits<ordinal_type>::max();
+
     // count the union of nonzeros in Arow and Brow
     size_type numEntries = 0;
     size_type ai         = 0;
@@ -201,67 +201,6 @@ struct UnmergedSumFunctor {
   CcolindsT Ccolinds;
   CcolindsT ABperm;
 };
-
-template <typename ExecSpace, typename size_type, typename ordinal_type,
-          typename CrowptrsT, typename CcolindsT>
-struct SortEntriesFunctor {
-  SortEntriesFunctor(const CrowptrsT& Crowptrs_, const CcolindsT& Ccolinds_,
-                     const CcolindsT& ABperm_)
-      : Crowptrs(Crowptrs_),
-        Ccolinds(Ccolinds_),
-        CcolindsAux("C colind aux", Ccolinds_.extent(0)),
-        ABperm(ABperm_),
-        ABpermAux("AB perm aux", ABperm_.extent(0)) {}
-  typedef typename Kokkos::TeamPolicy<ExecSpace>::member_type TeamMember;
-  KOKKOS_INLINE_FUNCTION void operator()(const TeamMember t) const {
-    // 3: Sort each row's colinds (permuting values at same time), then count
-    // unique colinds (write that to Crowptr(i)) CrowptrTemp tells how many
-    // entries in each oversized row
-    ordinal_type i       = t.league_rank();
-    size_type rowStart   = Crowptrs(i);
-    size_type rowEnd     = Crowptrs(i + 1);
-    size_type rowNum     = rowEnd - rowStart;
-    using lno_t          = typename CcolindsT::non_const_value_type;
-    using unsigned_lno_t = typename std::make_unsigned<lno_t>::type;
-    KokkosKernels::Impl::SerialRadixSort2<size_type, unsigned_lno_t, lno_t>(
-        (unsigned_lno_t*)Ccolinds.data() + rowStart,
-        (unsigned_lno_t*)CcolindsAux.data() + rowStart,
-        ABperm.data() + rowStart, ABpermAux.data() + rowStart, rowNum);
-  }
-  CrowptrsT Crowptrs;
-  CcolindsT Ccolinds;
-  CcolindsT CcolindsAux;
-  CcolindsT ABperm;
-  CcolindsT ABpermAux;
-};
-
-#ifdef KOKKOS_ENABLE_CUDA
-template <typename size_type, typename ordinal_type, typename CrowptrsT,
-          typename CcolindsT>
-struct SortEntriesFunctor<Kokkos::Cuda, size_type, ordinal_type, CrowptrsT,
-                          CcolindsT> {
-  SortEntriesFunctor(const CrowptrsT& Crowptrs_, CcolindsT& Ccolinds_,
-                     CcolindsT& ABperm_)
-      : Crowptrs(Crowptrs_), Ccolinds(Ccolinds_), ABperm(ABperm_) {}
-  typedef typename Kokkos::TeamPolicy<Kokkos::Cuda>::member_type TeamMember;
-  KOKKOS_INLINE_FUNCTION void operator()(const TeamMember t) const {
-    // 3: Sort each row's colinds (permuting values at same time), then count
-    // unique colinds (write that to Crowptr(i)) CrowptrTemp tells how many
-    // entries in each oversized row
-    size_type i        = t.league_rank();
-    size_type rowStart = Crowptrs(i);
-    size_type rowEnd   = Crowptrs(i + 1);
-    size_type rowNum   = rowEnd - rowStart;
-    KokkosKernels::Impl::TeamBitonicSort2<
-        size_type, typename CcolindsT::non_const_value_type,
-        typename CcolindsT::non_const_value_type, TeamMember>(
-        Ccolinds.data() + rowStart, ABperm.data() + rowStart, rowNum, t);
-  }
-  CrowptrsT Crowptrs;
-  CcolindsT Ccolinds;
-  CcolindsT ABperm;
-};
-#endif
 
 template <typename size_type, typename ordinal_type, typename ArowptrsT,
           typename BrowptrsT, typename CrowptrsT, typename CcolindsT>
@@ -478,7 +417,6 @@ template <typename size_type, typename ordinal_type, typename ArowptrsT,
           typename BscalarT>
 struct SortedNumericSumFunctor {
   using CscalarT = typename CvaluesT::non_const_value_type;
-  static constexpr ordinal_type ORDINAL_MAX = std::numeric_limits<ordinal_type>::max();
 
   SortedNumericSumFunctor(const ArowptrsT& Arowptrs_,
                           const BrowptrsT& Browptrs_,
@@ -502,6 +440,8 @@ struct SortedNumericSumFunctor {
 
   KOKKOS_INLINE_FUNCTION void operator()(const ordinal_type i) const
   {
+    const ordinal_type ORDINAL_MAX = Kokkos::ArithTraits<ordinal_type>::max();
+
     // count the union of nonzeros in Arow and Brow
     size_type ai         = 0;
     size_type bi         = 0;

